src/Storages/MaterializedView/RefreshSet.cpp

```diff
@@ -0,0 +1,141 @@
+#include <Storages/MaterializedView/RefreshSet.h>
+#include <Storages/MaterializedView/RefreshTask.h>
+
+namespace CurrentMetrics
+{
+    extern const Metric RefreshableViews;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+RefreshSet::Handle::Handle(Handle && other) noexcept
+{
+    *this = std::move(other);
+}
+
+RefreshSet::Handle & RefreshSet::Handle::operator=(Handle && other) noexcept
+{
+    if (this == &other)
+        return *this;
+    reset();
+    parent_set = std::exchange(other.parent_set, nullptr);
+    id = std::move(other.id);
+    dependencies = std::move(other.dependencies);
+    metric_increment = std::move(other.metric_increment);
+    return *this;
+}
+
+RefreshSet::Handle::~Handle()
+{
+    reset();
```

Comment:
Hm, is this a thing in practice? Do we handle it in other places? What handling would be useful? std::abort()?

I was assuming that valid mutex calls don't fail unless the OS is in such a bad state (can't allocate a few bytes for the futex waiter state?) that it doesn't matter what we do.

Reply:
My comment was specific to calling a function that throws from a noexcept function (such as d-tors since c++11).
std::terminate(defaulting to std::abort) will be called at the moment a noexcept function throws.

I believe reset() can be placed within a try-catch block with exception logging. This way, in case of an exception, the stack unwinds, and destructors of local objects are called.

Given that, in this case, the exception may be caused by the underlying OS and not by the application code, it's okay to leave it as you say.